eclair-core/src/main/scala/fr/acinq/eclair/swap/SwapHelpers.scala

```diff
@@ -0,0 +1,133 @@
+import fr.acinq.eclair.payment.send.PaymentInitiator.SendPaymentToNode
+import fr.acinq.eclair.payment.{Bolt11Invoice, PaymentEvent}
+import fr.acinq.eclair.swap.SwapCommands._
+import fr.acinq.eclair.swap.SwapEvents.TransactionPublished
+import fr.acinq.eclair.swap.SwapTransactions.makeSwapOpeningTxOut
+import fr.acinq.eclair.transactions.Transactions.{TransactionWithInputInfo, checkSpendable}
+import fr.acinq.eclair.wire.protocol.{HasSwapId, OpeningTxBroadcasted, SwapInAgreement, SwapInRequest}
+import fr.acinq.eclair.{NodeParams, ShortChannelId}
+import scodec.bits.ByteVector
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.reflect.ClassTag
+import scala.util.{Failure, Success}
+
+object SwapHelpers {
+
+  def receiveSwapMessage[B <: SwapCommand : ClassTag](context: ActorContext[SwapCommand], stateName: String)(f: B => Behavior[SwapCommand]): Behavior[SwapCommand] =
+    Behaviors.receiveMessage {
+      case m: B => f(m)
+      case m =>
+        context.log.error(s"received unhandled message in ${context.self.toString} at state $stateName of ${m.getClass.getSimpleName}")
+        Behaviors.same
+    }
+
+  def swapInvoiceExpiredTimer(swapId: String): String = "swap-invoice-expired-timer-" + swapId
+
+  def swapOpeningTimer(swapId: String): String = "swap-opening-timer-" + swapId
+
+  def watchForTxConfirmation(watcher: ActorRef[ZmqWatcher.Command])(replyTo: ActorRef[WatchTxConfirmedTriggered], txId: ByteVector32, minDepth: Long): Unit =
+    watcher ! WatchTxConfirmed(replyTo, txId, minDepth)
+
+  def queryChannelInfoShortId(register: actor.ActorRef, shortChannelId: ShortChannelId)(implicit context: ActorContext[SwapCommand]): Unit =
+    register ! Register.ForwardShortId(getInfoAdapter(context).toClassic, shortChannelId, CMD_GET_CHANNEL_INFO(getInfoAdapter(context).toClassic))
+
+  def queryChannelInfo(register: actor.ActorRef, channelId: ByteVector32)(implicit context: ActorContext[SwapCommand]): Unit =
+    register ! Register.Forward(getInfoAdapter(context).toClassic, channelId, CMD_GET_CHANNEL_INFO(getInfoAdapter(context).toClassic))
+
+  def getInfoAdapter(context: ActorContext[SwapCommand]): ActorRef[CommandSuccess[CMD_GET_CHANNEL_INFO]] = context.messageAdapter[CommandSuccess[CMD_GET_CHANNEL_INFO]](ChannelInfoResponse)
+
+  def queryOnChainBalance(wallet: OnChainWallet)(implicit context: ActorContext[SwapCommand]): Unit =
+    context.pipeToSelf(wallet.onChainBalance())(QueryOnChainBalance)
+
+  def payInvoice(nodeParams: NodeParams)(paymentInitiator: actor.ActorRef, swapId: String, invoice: Bolt11Invoice): Unit =
+    paymentInitiator ! SendPaymentToNode(invoice.amount_opt.get, invoice, nodeParams.maxPaymentAttempts, Some(swapId), assistedRoutes = invoice.routingInfo, nodeParams.routerConf.pathFindingExperimentConf.getRandomConf().getDefaultRouteParams, blockUntilComplete = true)
+
+  def watchForPayment(watch: Boolean)(implicit context: ActorContext[SwapCommand]): Unit =
```

Comment:
This is a good idea! But I think we should be more specific instead of registering to `PaymentEvent`. We should explicitly register only to `PaymentReceived` when we generated an invoice and wait for it to be paid, and we should only register to `PaymentSent` and `PaymentFailed` when we're paying an invoice.

Reply:
Fixed in [ad33a79](https://github.com/ACINQ/eclair/pull/2496/commits/ad33a79074fc1151abea0820d0cc7c67821fb15d). This also forced some reorganization of the coded that handles resume/restart scenarios because a Taker actor check points right before trying to pay the invoice. 

Now the `payInvoice` method first checks if the payment is in the database before paying the invoice. This ensures the `Taker` will not be pay the fee or claim invoices twice after a restart or resume.